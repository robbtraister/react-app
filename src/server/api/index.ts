import express from 'express'

const router = express()

router.use((req, res, next) => {
  res.sendStatus(404)
})

export default router
