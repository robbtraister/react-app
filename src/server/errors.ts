// import express from 'express'

export class RedirectError extends Error {
  location: string
  statusCode: number

  constructor(location: string, statusCode = 302) {
    super(`Redirect: ${location}`)
    this.location = location
    this.statusCode = statusCode
  }
}

export default [
  (err, req, res, next) => {
    if (
      err.location &&
      err.statusCode &&
      err.statusCode >= 300 &&
      err.statusCode < 400
    ) {
      res.redirect(err.location)
    } else {
      next(err)
    }
  },
  (_err, req, res, next) => {
    res.sendStatus(500)
  }
]
