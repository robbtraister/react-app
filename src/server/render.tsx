import fs from 'fs'
import path from 'path'

import React from 'react'
import ReactDOM from 'react-dom/server'
import { StaticRouter } from 'react-router'

import { RedirectError } from './errors'

import { App } from '~/app'
import pages from '~/app/pages'

export async function render({ format = 'html', location }) {
  const page =
    location
      .replace(/[?&#].+/, '')
      .replace(/^\/+/, '')
      .split('/')[0] || 'dashboard'

  const [{ default: Format }] = await Promise.all([
    import(`../formats/${format}`)
  ])

  const style =
    format === 'amp'
      ? (
          await fs.promises.readFile(
            // eslint-disable-next-line no-eval
            path.join('.', 'build', 'dist', `styles.css`)
          )
        ).toString()
      : null

  const context: { url?: string } = {}

  const html = ReactDOM.renderToString(
    <StaticRouter context={context} location={location}>
      <Format page={page} style={style}>
        <App pages={pages} />
      </Format>
    </StaticRouter>
  )

  if (context.url) {
    throw new RedirectError(context.url)
  }

  return html
}

export default async function (req, res, next) {
  try {
    res.send(
      `<!DOCTYPE html>${await render({
        location: req.originalUrl,
        format: req.query.format
      })}`
    )
  } catch (error) {
    next(error)
  }
}
