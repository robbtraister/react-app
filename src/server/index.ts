import express from 'express'

import api from './api'
import render from './render'
import errors from './errors'

export function app() {
  const app = express()

  app.disable('x-powered-by')

  app.use('/api', api)
  app.use(render)
  app.use(errors)

  return app
}

export function server(port = undefined) {
  const server = app()
  server.listen(Number(port || process.env.PORT) || 3000, err => {
    err ? console.error(err) : console.log(`Listening at: ${server.address()}`)
  })
}

export default server

// eslint-disable-next-line no-eval
if (eval('module === require.main')) {
  server()
}
