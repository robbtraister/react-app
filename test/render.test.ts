import request from 'supertest'

import { app } from '~/server'

describe('Render', () => {
  it('redirect root', () => {
    return request(app()).get('/').expect(302).expect('Location', '/home')
  })

  it('render home', () => {
    return request(app())
      .get('/home')
      .expect('Content-Type', /^text\/html(;|$)/)
      .then(response => {
        expect(response.text).toMatchSnapshot()
      })
  })

  it('render home in amp', () => {
    return request(app())
      .get('/home?format=amp')
      .expect('Content-Type', /^text\/html(;|$)/)
      .then(response => {
        expect(response.text).toMatchSnapshot()
      })
  })
})
