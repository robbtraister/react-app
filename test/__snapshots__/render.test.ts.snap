// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render render home 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><base href=\\"/\\"/><meta charSet=\\"utf-8\\"/><meta name=\\"viewport\\" content=\\"width=device-width,minimum-scale=1,initial-scale=1\\"/><link rel=\\"stylesheet\\" href=\\"dist/styles.css\\"/><script src=\\"dist/html.js\\" defer=\\"\\"></script></head><body><div id=\\"root\\"><h1>Home</h1></div></body></html>"`;

exports[`Render render home in amp 1`] = `
"<!DOCTYPE html><html amp=\\"\\" lang=\\"en\\"><head><base href=\\"/\\"/><link rel=\\"canonical\\" href=\\"https://localhost:8080/home?format=amp\\"/><meta charSet=\\"utf-8\\"/><meta name=\\"viewport\\" content=\\"width=device-width,minimum-scale=1,initial-scale=1\\"/><script async=\\"\\" src=\\"https://cdn.ampproject.org/v0.js\\"></script><script async=\\"\\" custom-element=\\"amp-script\\" src=\\"https://cdn.ampproject.org/v0/amp-script-0.1.js\\"></script><style amp-boilerplate=\\"\\">body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate=\\"\\">body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript><style amp-custom=\\"\\">html {
  height: 100%;
}

body {
  height: 100%;
  margin: 0;
}

a {
  text-decoration: none;
}

#root {
  display: flex;
  flex-direction: column;
  height: 100%;
}
</style></head><body><amp-script layout=\\"responsive\\" width=\\"9\\" height=\\"16\\" src=\\"https://localhost:8080/dist/amp.js\\" sandbox=\\"allow-forms\\"><input type=\\"hidden\\" name=\\"location\\" value=\\"/home?format=amp\\"/><div id=\\"root\\"></div></amp-script></body></html>"
`;
